### PR TITLE
docs: introduce TypeScript documentation

### PR DESCRIPTION
**Pull request by wellwelwel**
I added an installation case for `@types/node` in **README.md**, but described its purpose in `/documentation/en/TypeScript-Examples.md`.

Every example in this documentation was tested 🤹🏻‍♀️

> You can follow the documentation [here](https://github.com/wellwelwel/node-mysql2/blob/ts-docs/documentation/en/TypeScript-Examples.md).

---

### Annotations:
- [x] Create practical examples based on Issues in `examples/typescript/`
- [x] Check spelling after completing the documentation

> Every file that was changed has a direct relation to the documentation 🙋🏻‍♂️

### CODE REVIEW

**Comment by wellwelwel** [comment]
@sidorares

After several tests, I never got an `OkPacket` response.
Also searching through the code, all the uses of `OK` that I found, are only used internally.

Instead, every query always returns `RowDataPacket[]`, `RowDataPacket[][]`, `ResultSetHeader` or `ResultSetHeader[]`.
> Including `query().on('result')`

<details>
<summary>
<b>Examples from source code</b>
</summary>

- Internal usage:
https://github.com/sidorares/node-mysql2/blob/8ca36c774a526676f114883842d896f21e20f698/lib/connection.js#L882-L887
  - > `writeOk` returns a *void*

- `query`:
  - https://github.com/sidorares/node-mysql2/blob/8ca36c774a526676f114883842d896f21e20f698/lib/commands/query.js#L117
  - > only exports `row` and `ResultSetHeader`

- `execute`:
https://github.com/sidorares/node-mysql2/blob/8ca36c774a526676f114883842d896f21e20f698/lib/commands/execute.js#L98
  - > only exports `row` and `ResultSetHeader`
  
</details>

---

### Handling it considering not to cause a break change
1. Set `changedRows` in `ResultSetHeader` as a required property

- **JS**
  - In `/lib/packets/resultset_header.js`:
  ```diff
    if (m !== null) { 
      this.changedRows = parseInt(m[1], 10); 
  + } else {
  +   this.changedRows = 0;
    }
  ```
- **TS**
  - In `typings/mysql/lib/protocol/packets/ResultSetHeader.d.ts`:
  ```diff
  -   changedRows?: number;
  +   changedRows: number;
    }

    export { ResultSetHeader };
  ```

2. Add `ResultSetHeader[]` possibility to `query` and `execute` (for `multipleStatements`)
3. **Optional**: comment in `typings/mysql/lib/protocol/packets/OkPacket.d.ts`
    ```diff
    + /** @deprecated */
      declare interface OkPacket {
    ```
    - >Later, in a major version, remove the `OkPacket` and `OkPacket[]` from `query` and `execute` types.

---

That said, I don't propose to use the `OkPacket` type in TypeScript documentation.

@sidorares, I would like to know your feedback on this question.
For now, I will continue the TypeScript documentation, but ignoring the `OkPacket`.

**Comment by wellwelwel** [comment]
@sidorares, the documentation itself is done, in case you want to [read it](https://github.com/wellwelwel/node-mysql2/blob/ts-docs/documentation/en/TypeScript-Examples.md) 🕺🏻

I will create four advanced examples based on **Issues**, then review it before completing this **PR**.

File: examples/typescript/row-data-packet.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+interface User extends RowDataPacket {
+  id: number;
+  name: string;
+}
+
+(async () => {
+  const access: ConnectionOptions = {
+    host: 'localhost',
+    user: 'root',
```
[the remark this answers is not in the document]
Comment (wellwelwel) [reply examples/typescript/row-data-packet.ts]: 🤨

**Comment by wellwelwel** [comment]
The documentation and the examples are ready 🕺🏻

I have created seven "*advanced*" examples based on **Issues**.
Tomorrow, I will review them before completing this **PR**.

**Comment by sidorares** [comment]
Thanks @wellwelwel!

I think the naming confusion comes from the fact it's called "OK_Packet" in the [documentation](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_ok_packet.html) and it's a [ResultSetHeader](https://github.com/sidorares/node-mysql2/blob/00508c38a13f1394fb8a6cf3fd88df1353ba0e77/lib/packets/resultset_header.js#L13) here. We should probably only use ResultSetHeader name for consistency

`changedRows` is not defined at all in the protocol, but the property existed in `mysqljs/mysql` so we support it as well. It's basically just a "view" into a human-readable "info" part if the info contains "changed [number]" in it's text. I'm not sure if we should set it to 0 if it can't be parsed - wdyt @wellwelwel ?

**Comment by wellwelwel** [comment]
Hey, @sidorares 🙋🏻‍♂️

I have chosen `0` to not cause a break change.
I did a quick test and even a personal project would break if I changed from `number` to another type, like `false` or `null`, for example.

About keep it as it is...

I explain it better in my previous comment and complementing, the reason for this is to pattern the main difference between `OkPacket` and `ResultSetHeader`, especially to be able to explain the ResultSetHeader and its usability.

Feel free to change it, I'm always open for ideas and suggestions 🤹🏻‍♀️

Just complementing (again), basically the `changedRows` appears in `ResultSetHeader` when an `UPDATE` query is performed.

**Comment by wellwelwel** [comment]
For optional fields, especially in TypeScript, is recommended check if the fields exists, and if yes, the `changedRows` expects for a `number`.

- It would work perfectly, because it check for type:
```js
typeof myResultSetHeader?.changedRows === 'number';
```

- Causing a break change:
```js
'changedRows' in myResultSetHeader && myResultSetHeader.changedRows > 0;
```

**Pull request merged**